File: parsers/dart/test/diff_test.dart
import 'package:synx/synx.dart';
import 'package:test/test.dart';

void main() {
  test('identical', () {
    final a = SynxObject()
      ..set('x', synxInt(1))
      ..set('y', synxInt(2));
    final b = SynxObject()
      ..set('x', synxInt(1))
      ..set('y', synxInt(2));
    final d = diffObjects(a, b);
    expect(d.added.isEmpty, isTrue);
    expect(d.removed.isEmpty, isTrue);
    expect(d.changed, isEmpty);
    expect(d.unchanged.length, equals(2));
  });

  test('added/removed', () {
    final a = SynxObject()..set('x', synxInt(1));
    final b = SynxObject()..set('y', synxInt(2));
    final d = diffObjects(a, b);
    expect(d.added.length, equals(1));
    expect(d.removed.length, equals(1));
  });

  test('changed', () {
    final a = SynxObject()..set('name', synxString('Alice'));
    final b = SynxObject()..set('name', synxString('Bob'));
    final d = diffObjects(a, b);
    expect(d.changed.length, equals(1));
    expect(d.changed.first.key, equals('name'));
  });
}

File: parsers/go/diff_test.go
package synx

import "testing"

func TestDiffIdentical(t *testing.T) {
	a := NewObjectMap()
	a.Set("x", Int(1))
	a.Set("y", Int(2))
	b := NewObjectMap()
	b.Set("y", Int(2))
	b.Set("x", Int(1))
	d := Diff(a, b)
	if !d.Added.IsEmpty() || !d.Removed.IsEmpty() || len(d.Changed) != 0 {
		t.Fatalf("identical mismatch: %+v", d)
	}
	if len(d.Unchanged) != 2 {
		t.Fatalf("unchanged len: %d", len(d.Unchanged))
	}
}

func TestDiffAddedRemoved(t *testing.T) {
	a := NewObjectMap()
	a.Set("x", Int(1))
	b := NewObjectMap()
	b.Set("y", Int(2))
	d := Diff(a, b)
	if d.Added.Len() != 1 || d.Removed.Len() != 1 {
		t.Fatalf("added/removed mismatch: %+v", d)
	}
}

func TestDiffChanged(t *testing.T) {
	a := NewObjectMap()
	a.Set("name", String("Alice"))
	b := NewObjectMap()
	b.Set("name", String("Bob"))
	d := Diff(a, b)
	if len(d.Changed) != 1 || d.Changed[0].Key != "name" {
		t.Fatalf("changed mismatch: %+v", d)
	}
}
